/*
 * Copyright 2014 Stanford University.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * - Redistributions of source code must retain the above copyright
 *   notice, this list of conditions and the following disclaimer.
 *
 * - Redistributions in binary form must reproduce the above copyright
 *   notice, this list of conditions and the following disclaimer in the
 *   documentation and/or other materials provided with the
 *   distribution.
 *
 * - Neither the name of the copyright holders nor the names of
 *   its contributors may be used to endorse or promote products derived
 *   from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
 * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

 /*
  * Protocol buffer messages between workers and the scheduler.
  * Author: Philip Levis <pal@cs.stanford.edu>
  */

message LogicalIDSet {
  repeated uint64 ids = 1 [packed=true];
}

message PhysicalIDSet {
  repeated uint64 ids = 1 [packed=true];
}

message JobIDSet {
  repeated uint64 ids = 1 [packed=true];        
}

message SubmitComputeJobCommand {
  optional string name            = 1;
  optional uint64 job_id          = 2;
  optional LogicalIDSet read_set  = 3;
  optional LogicalIDSet write_set = 4;
  optional JobIDSet before_set    = 5;
  optional JobIDSet after_set     = 6;
  optional uint64 parent_id       = 7;
  optional uint64 future_id       = 8;
  optional bytes params           = 9;
}

message ExecuteComputeJobCommand {
  optional string name             = 1;
  optional uint64 job_id           = 2;
  optional PhysicalIDSet read_set  = 3;
  optional PhysicalIDSet write_set = 4;
  optional JobIDSet before_set     = 5;
  optional JobIDSet after_set      = 6;
  optional uint64 parent_id        = 7;
  optional uint64 future_id        = 8; 
  optional bytes params            = 9;
  
}
